File: edge/app/backup/fingerprint.py
from __future__ import annotations

import hashlib

from app.backup.filters import DiscoveredFile


def build_manifest(files: list[DiscoveredFile]) -> str:
    lines = [
        f"{file.archive_path}\t{file.size}"
        for file in sorted(files, key=lambda item: item.archive_path)
    ]
    return "\n".join(lines)


def compute_fingerprint(files: list[DiscoveredFile]) -> str:
    manifest = build_manifest(files)
    return hashlib.sha256(manifest.encode("utf-8")).hexdigest()

File: edge/internal/backup/fingerprint.go
package backup

import (
	"cmp"
	"crypto/sha256"
	"fmt"
	"slices"
	"strings"
)

// ComputeFingerprint returns a SHA-256 hex digest of the sorted path+size manifest.
func ComputeFingerprint(files []*DiscoveredFile) string {
	sorted := make([]*DiscoveredFile, len(files))
	copy(sorted, files)
	slices.SortFunc(sorted, func(a, b *DiscoveredFile) int {
		return cmp.Compare(a.ArchivePath, b.ArchivePath)
	})

	var lines []string
	for _, f := range sorted {
		lines = append(lines, fmt.Sprintf("%s\t%d", f.ArchivePath, f.Size))
	}
	manifest := strings.Join(lines, "\n")
	sum := sha256.Sum256([]byte(manifest))
	return fmt.Sprintf("%x", sum)
}
